File: dsa-problems/TreesDP/solutions/javascript/TDP-014-centroid-decomp-time-decay.js
const readline = require("readline");
const rl = readline.createInterface({ input: process.stdin, terminal: false });

const lines = [];
rl.on("line", (line) => lines.push(line.trim()));
rl.on("close", () => {
  let idx = 0;
  const [n, D] = lines[idx++].split(" ").map(Number);

  const adj = Array.from({ length: n + 1 }, () => []);
  for (let i = 0; i < n - 1; i++) {
    const [u, v, w] = lines[idx++].split(" ").map(Number);
    adj[u].push([v, w]);
    adj[v].push([u, w]);
  }

  const q = parseInt(lines[idx++]);
  const marked = new Map();
  const results = [];

  function bfs(start) {
    const dist = new Array(n + 1).fill(Infinity);
    dist[start] = 0;
    const queue = [start];
    let head = 0;
    while (head < queue.length) {
      const u = queue[head++];
      for (const [v, w] of adj[u]) {
        if (dist[u] + w < dist[v]) {
          dist[v] = dist[u] + w;
          queue.push(v);
        }
      }
    }
    let minCost = Infinity;
    for (const [node, val] of marked) {
      if (dist[node] !== Infinity) {
        minCost = Math.min(minCost, dist[node] + val);
      }
    }
    return minCost;
  }

  for (let i = 0; i < q; i++) {
    const parts = lines[idx++].split(" ").map(Number);
    if (parts[0] === 1) {
      const [_, v, val, t] = parts;
      marked.set(v, val);
    } else {
      const [_, v, t] = parts;
      results.push(bfs(v));
    }
  }

  console.log(results.join("\n"));
});

File: dsa-problems/TreesDP/solutions/python/TDP-014-centroid-decomp-time-decay.py
import sys
from collections import deque

def main():
    data = sys.stdin.read().split()
    idx = 0
    n = int(data[idx]); idx += 1
    D = int(data[idx]); idx += 1

    adj = [[] for _ in range(n + 1)]
    for _ in range(n - 1):
        u, v, w = int(data[idx]), int(data[idx + 1]), int(data[idx + 2])
        idx += 3
        adj[u].append((v, w))
        adj[v].append((u, w))

    q = int(data[idx]); idx += 1
    marked = {}
    results = []

    def bfs(start):
        dist = [float('inf')] * (n + 1)
        dist[start] = 0
        queue = deque([start])
        while queue:
            u = queue.popleft()
            for v, w in adj[u]:
                if dist[u] + w < dist[v]:
                    dist[v] = dist[u] + w
                    queue.append(v)
        min_cost = float('inf')
        for node, val in marked.items():
            if dist[node] != float('inf'):
                min_cost = min(min_cost, dist[node] + val)
        return min_cost

    for _ in range(q):
        qtype = int(data[idx]); idx += 1
        if qtype == 1:
            v = int(data[idx]); idx += 1
            val = int(data[idx]); idx += 1
            t = int(data[idx]); idx += 1
            marked[v] = val
        else:
            v = int(data[idx]); idx += 1
            t = int(data[idx]); idx += 1
            results.append(str(bfs(v)))

    print('\n'.join(results))

if __name__ == "__main__":
    main()
